datapath: Devices table and controller

joamaki opened the pull request:
This PR adds the following new tables to pkg/datapath/tables:
```
// Device is a local network device along with addresses associated with it.
type Device struct {
	Index        int              // positive integer that starts at one, zero is never used
	MTU          int              // maximum transmission unit
	Name         string           // e.g., "en0", "lo0", "eth0.100"
	HardwareAddr net.HardwareAddr // IEEE MAC-48, EUI-48 and EUI-64 form
	Flags        net.Flags        // e.g. net.FlagUp, net.eFlagLoopback, net.FlagMulticast

	Selected    bool            // If true this device can be used by Cilium
	Addrs       []DeviceAddress // Addresses assigned to the device
	RawFlags    uint32          // Raw interface flags
	Type        string          // Device type, e.g. "veth" etc.
	MasterIndex int             // Index of the master device (e.g. bridge or bonding device)
}

var deviceTableSchema = &memdb.TableSchema{
	Name: "devices",
	Indexes: map[string]*memdb.IndexSchema{
		string(statedb.IDIndex): {
			Name:         string(statedb.IDIndex),
			AllowMissing: false,
			Unique:       true,
			Indexer:      &memdb.IntFieldIndex{Field: "Index"},
		},
		string(DeviceNameIndex): {
			Name: string(DeviceNameIndex),
			// Name can be temporarily missing if we create the device
			// from an address update.
			AllowMissing: true,
			Unique:       true,
			Indexer:      &memdb.StringFieldIndex{Field: "Name"},
		},
	},
}

type Route struct {
	Table     int
	LinkIndex int

	Scope uint8
	Dst   netip.Prefix
	Src   netip.Addr
	Gw    netip.Addr
}

var (
	routeTableSchema = &memdb.TableSchema{
		Name: "routes",
		Indexes: map[string]*memdb.IndexSchema{
			"id": {
				Name:         "id",
				AllowMissing: false,
				Unique:       true,
				Indexer: &memdb.CompoundIndex{
					Indexes: []memdb.Indexer{
						&memdb.IntFieldIndex{Field: "Table"},
						&memdb.IntFieldIndex{Field: "LinkIndex"},
						&statedb.NetIPPrefixFieldIndex{Field: "Dst"},
					},
				},
			},
			string(linkIndexIndex): {
				Name:         string(linkIndexIndex),
				AllowMissing: false,
				Unique:       false,
				Indexer:      &memdb.IntFieldIndex{Field: "LinkIndex"},
			}},
	}
)

```

The tables are populated by the DevicesController by subscribing to links, addresses and routes using netlink.
To avoid large changes to existing code, this PR keeps the existing DeviceManager API and validates the
new implementation using the old existing test suite.

Follow-up PRs will incrementally switch uses of `option.Config.GetDevices()` to querying and watching the devices table.

joamaki commented on pkg/option/config.go:
```diff
@@ -2498,12 +2498,6 @@ func (c *DaemonConfig) TunnelExists() bool {
 	return c.TunnelingEnabled() || c.EnableIPv4EgressGateway
 }
 
-// AreDevicesRequired returns true if the agent needs to attach to the native
```
It seems likely that we can drop this construct and just always detect devices. Will need to think it through...

joamaki committed:
datapath/tables: Add device and route state tables

Add state tables for storing network devices (along with addresses)
and routes.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
datapath/linux: Implement DevicesController

DevicesController populates the devices and routes tables.

To avoid large refactoring of the existing code around option.Config.GetDevices()
this opts for an incremental approach: keep the DeviceManager API as-is and reuse
its test to validate that the existing functionality does not break.

The tests required some slight adaptation: the "DeviceManager" is re-created after
each test case to recreate the initial devices table.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on pkg/datapath/linux/devices.go:
```diff
@@ -1,521 +1,125 @@
-				return false
+	var nodeDevice *state.Device
+	if k8sEnabled {
+		nodeIP := node.GetK8sNodeIP()
```
This requires use of LocalNodeStore and changes in https://github.com/cilium/cilium/pull/24866 to work as NodeIP is otherwise uninitialized.

joamaki replied:
#24866 has been merged.

brb commented on pkg/datapath/state/device.go:
```diff
@@ -0,0 +1,137 @@
+
+// NodePortAddr returns the NodePort frontend addresses chosen
+// from this device.
+func (dev *Device) NodePortAddr() NodePortAddr {
```
Do we actually need this function to be NodePort-specific? I guess in the future we will support multiple NodePort IP addrs per netdev, so maybe we could use a generic `IPAddrs()` method instead?

brb replied:
To elaborate more - I don't want to leak any LB details (in this case NodePort) into this relatively generic package.

joamaki replied:
Since this method is about picking the single nodeport frontend for the device I would keep it NodePort-specific. However, you're right it shouldn't be here. I'll move it to a more appropriate place. In the future this would go away completely once service load-balancing is refactored to support multiple frontends per device.

tommyp1ckles commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,527 @@
+		Namespace:    dc.NetNS,
+		ListExisting: false,
+	}
+	if err := netlink.AddrSubscribeWithOptions(addrUpdates, dc.ctx.Done(), opts); err != nil {
```
I believe in the past we've seen issues where the netlink can be unreliabe, should we try to reestablish the nl connection/sub if the channel gets closed?

joamaki replied:
Oh that's a bummer. Would require fair bit of additional complexity to resubscribe as it needs to relist from scratch in order to make sure no update has been missed.

Do you have any more info about the unreliability?

joamaki replied:
netlink man page has this to say: 
```
       Netlink is not a reliable protocol.  It tries its best to deliver
       a message to its destination(s), but may drop messages when an
       out-of-memory condition or other error occurs.  For reliable
       transfer the sender can request an acknowledgement from the
       receiver by setting the NLM_F_ACK flag.  An acknowledgement is an
       NLMSG_ERROR packet with the error field set to 0.  The
       application must generate acknowledgements for received messages
       itself.  The kernel tries to send an NLMSG_ERROR message for
       every failed packet.  A user process should follow this
       convention too.

       However, reliable transmissions from kernel to user are
       impossible in any case.  The kernel can't send a netlink message
       if the socket buffer is full: the message will be dropped and the
       kernel and the user-space process will no longer have the same
       view of kernel state.  It is up to the application to detect when
       this happens (via the ENOBUFS error returned by [recvmsg(2)](https://man7.org/linux/man-pages/man2/recvmsg.2.html)) and
       resynchronize.
```

The go `netlink` library doesn't do the ACKs so this is indeed is pretty unreliable and may miss updates.
Not sure in what scenarios the netlink socket would get closed, but wouldn't be surprised it can happen
in OOM situations.

I'll need to think a bit how we can make this reliable...

EDIT: Could maybe check sequence numbers of all updates and then restart everything from scratch if there's a missing update or any other error. Though maybe `ENOBUFS` would be sent to the `ErrorCallback` when that would happen.

tommyp1ckles replied:
Yeah, correct.

A timeout was added by a contributor to the[ nl library](https://github.com/cilium/cilium/issues/14710) to solve issues where the socket would hang indefinitely (I don't think it was clear what the actual underlying issue was, potentially just a bug in netlink?).

Could we propose/add the feature ourselves?

joamaki replied:
The recvfrom hang has been fixed 1 year ago: https://github.com/vishvananda/netlink/pull/793. Unfortunately the netlink library seems to not be maintained well anymore. I'll likely look into actively maintained alternative libraries that would also allow cleanly distinguishing between initial listing and updates.

joamaki committed:
statedb: Rename IPIndexer to IPFieldIndex, add NetIPPrefixFieldIndex

For consistency rename IPIndexer to IPFieldIndex.
Add NetIPPrefixFieldIndex to allow indexing netip.Prefix.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

lmb commented on pkg/statedb/index.go:
```diff
@@ -74,6 +74,55 @@ func (ii *IPIndexer) PrefixFromArgs(args ...interface{}) ([]byte, error) {
+	Field string
+}
+
+func (i *IPNetFieldIndex) FromObject(obj interface{}) (bool, []byte, error) {
```
Could this have some tests? reflect code is notoriously hard to get right.

joamaki replied:
With the switch to netip I now added `NetIPPrefixFieldIndex` there and added tests for it.

lmb commented on pkg/datapath/tables/route.go:
```diff
@@ -0,0 +1,120 @@
+	LinkName  string
+	LinkIndex int
+
+	Scope uint8
```
Is this the same scope as `DeviceAddress.Scope`? If yes, why is it `uint8` here? And would it be worth to make a new type for `Scope`?

tommyp1ckles replied:
fwiw the netlink library uses an alias type of uint8 to represent scope.

joamaki replied:
Using `uint8` now in route and address. `netlink` library for some reason used `int` as scope in the address.

lmb commented on daemon/cmd/daemon_main.go:
```diff
@@ -1629,6 +1629,7 @@ type daemonParams struct {
 	L2Announcer          *l2announcer.L2Announcer
 	L7Proxy              *proxy.Proxy
 	DB                   statedb.DB
+	DeviceManager        *linuxdatapath.DeviceManager `optional:"true"`
```
Oooh, daring! Why is it okay to have optional here? Does it make sense to document the assumptions?

joamaki replied:
This was pure laziness on my part as the plan was to immediately follow up with removal of `DeviceManager`. I'll look into doing this cleanly by making these part of the `Datapath` interface instead and including implementations in `FakeDatapath`.

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,531 @@
+
+	// Subscribe to address updates to find out about link address changes
+	// to update the device's addresses.
+	addrUpdates := make(chan netlink.AddrUpdate, 16)
```
Why 16? Seems like a magic number to me.

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,531 @@
+			if !timer.Stop() {
+				<-timer.C
+			}
+			for range addrUpdates {
```
Why does this need to drain the channels?

joamaki replied:
These channels need to be drained to release the netlink subscriptions. I'll add a comment.

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,531 @@
+// commits it to the state it returns true, if the batch is incomplete (e.g. route updates
+// for non-existing link), it returns false and the batch should be expanded with newer
+// updates and tried again.
+func (dc *devicesController) processBatch(batch map[int][]any) (ok bool) {
```
Maybe this should mention that batch is modified?

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,531 @@
+			// Update for a link we have not yet observed, abort and try again
+			// later.
+			dc.Log.Debugf("link %d not found, retrying later", index)
+			txn.Abort()
```
Why does one failed link update cancel the whole batch? Doesn't it make more sense to treat each link as individual units?

lmb replied:
Also, we might have mutated `dc.links` already at this point. That change doesn't need to be rolled back?

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,531 @@
+		}
+		// Sort the addresses so that secondary addresses are at bottom and then
+		// sorted by scope.
+		sort.SliceStable(d.Addrs, func(i, j int) bool {
```
I'm guessing that the order is important, but why? Could you add a test that the sort produces the correct output?

joamaki replied:
This sorting is only needed to pick the address for kube-proxy replacement, but that's not yet implemented. I removed the sorting from here and will add it directly to the function that picks the NodePort frontend address (once KPR gets ported to use devices table instead of the pkg/node globals).

lmb commented on pkg/datapath/linux/devices_controller_test.go:
```diff
@@ -0,0 +1,291 @@
+func devicesControllerTestSetup(t *testing.T) func() {
+	logging.SetLogLevelToDebug()
+	// Return the defer function.
+	return func() {
```
Why not use `t.Cleanup`?

joamaki replied:
Forgot it existed. Using it now.

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -76,117 +82,192 @@ type DevicesConfig struct {
+
+	// Avoid tight restart loop with an exponential backoff
+	backoff := backoff.Exponential{
+		Min:  100 * time.Millisecond,
```
What is the max? Is there a period after which the backoff resets to min?

joamaki replied:
Removed backoff completely as there isn't really a signal to reset. It's now using a static one second wait time between restarts.

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -40,18 +42,22 @@ var DevicesControllerCell = cell.Module(
 
+	// netlinkBuffer is the number of updates to buffer
+	// into the update channels.
+	netlinkBuffer = 16
```
Why 16 specifically? A bit of a joke, but I think in Go there are two valid constant channel sizes: 0 and 1 :D Otherwise you need to think about what you are buffering for and try and size this dynamically.

joamaki replied:
I'll drop the buffering as it's premature optimization.

tommyp1ckles commented on pkg/datapath/tables/device.go:
```diff
@@ -0,0 +1,115 @@
+	iter, err := r.Get(statedb.Query{Index: DeviceNameIndex})
+	if err != nil {
+		// table schema is malformed?
+		panic(err)
```
Should we be using log.Fatal instead of panic in these cases?

There also seems to be a lot of different failure paths for the statedb implementation of Get, seems potentially brittle to crash out here?

joamaki replied:
Yeah, that's the unfortunate effect of the very dynamic indexer functions in go-memdb: they reflect on the arguments and potentially accept many different forms. The failure in Get() is only if something unexpected is passed to the indexers `FromArgs` or an unknown index is used. 

My thinking here was that since this file both defines what the indexes are and this function, it is fine for this function to make the assumption that it does construct a correct query and thus panicing is fine. Any bug here would be triggered in tests anyway.

For the general case it'd be great to find a good way to construct queries for statedb tables in a safe way that cannot cause errors, but I'm not quite sure how to do that in an ergonomic way without getting overly fancy with generic types.

tommyp1ckles commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,531 @@
+
+	// Subscribe to address updates to find out about link address changes
+	// to update the device's addresses.
+	addrUpdates := make(chan netlink.AddrUpdate, 16)
```
why 16?

joamaki replied:
It's just a small amount of buffering to avoid unnecessary switches between the goroutine doing `recvfrom` the netlink socket and this one batching them up. The actual amount doesn't matter much. The buffering amount is now lifted into a constant defined at the top.

tommyp1ckles commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,531 @@
+		case <-dc.ctx.Done():
+			if !timer.Stop() {
+				<-timer.C
+			}
```
Q: Would it make sense to process any remaining batch here?

joamaki replied:
Not really, we're either stopping (in which case we don't care) or restarting (in which case we'll anyway flush the state and reinitialize).

tommyp1ckles commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,690 @@
+		// previous failed processing loop. We're flushing it here
+		// in the same write transaction as the updates so that observers
+		// do not see the tables emptying.
+		dc.flushState(txn)
```
Should we log or handle this error somehow?

What's the implication if these tables have leftover rows?

joamaki replied:
Yeah, could log the error in `flushState`. Or might even be better to `Fatal` since failure in `DeleteAll` means we've messed up the query/indexes. Will fix this.

EDIT: Doesn't even make sense to check the error. `DeleteAll` uses the "empty" `All` query that cannot fail. I'll simplify this.

tommyp1ckles commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,690 @@
+	}
+
+	for index, updates := range batch {
+		link := dc.links[index]
```
What happens if a link update occurs while this is going on?

joamaki replied:
There is no concurrent updates to `dc.links`. It only updates above in this same function. If a link update comes from the netlink socket then it'll be read into the next batch and be processed then.

joamaki replied:
`links` gone completely now so resolving.

joamaki commented on pkg/datapath/cells.go:
```diff
@@ -144,4 +153,9 @@ type datapathParams struct {
+	// Depend on DeviceManager to ensure devices have been resolved.
+	// This is required until option.Config.GetDevices() has been removed and
+	// uses of it converted to Table[Device].
+	DeviceManager *linuxdatapath.DeviceManager
```
This really annoys me. As long as we have globals like this we need to play dangerous games with the initialization ordering to get them initialized early enough. Next step is definitely the prompt removal of `option.Config.GetDevices` to make sure no new module accidentally accesses it uninitialized.

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,697 @@
+
+	// NetNS is an optional network namespace handle. Used by tests to restrict the detection
+	// to specific namespace.
+	NetNS *netns.NsHandle `optional:"true"`
```
Nice!

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,697 @@
+	initialized    chan struct{}
+	filter         deviceFilter
+	l3DevSupported bool
+	ctx            context.Context
```
I read somewhere that context shouldn't be stored in structs but passed as parameters. I guess we don't follow that in the cilium code base?

joamaki replied:
Yep, this can as well be passed as parameter. The cancel function is needed here for the stop hook. I'll refactor this to use `pkg/hive/job` instead which will do the context management for us. 

EDIT: Since netlink library hangs waiting for new message in `Recvfrom` even though we're stopping the subscriptions, it's a bit hard to use the `job` library. Instead I cleaned this up so that `ctx` is not stored in the `devicesController` struct, but just passed directly to `run`.

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,697 @@
+	// successfully.
+	select {
+	case <-dc.initialized:
+	case <-startCtx.Done():
```
What does it mean for startCtx to be cancelled? It doesn't tear down dc.run from what I can tell?

joamaki replied:
`startCtx` is cancelled on aborted start up (e.g. on SIGINT/SIGTERM). Good catch, the context should be cancelled to tear down dc.run.

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,697 @@
+
+func (dc *devicesController) initialize() error {
+	// Do initial listing for each address, routes and links. We cannot use
+	// 'ListExiting' option as it does not provide a mechanism to know when
```
```suggestion
	// 'ListExisting' option as it does not provide a mechanism to know when
```

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,697 @@
+	linkUpdates chan netlink.LinkUpdate,
+) {
+	// Collect all updates into a batch to minimize the
+	// number of write transactions when there are large
```
Writes to the statedb?

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,697 @@
+	}
+	timerStopped := true
+	defer func() {
+		if !timerStopped && !timer.Stop() {
```
Why is this necessary when timer is discarded anyways?

joamaki replied:
I refactored the code to just use a ticker instead. I don't think the optimization of resetting the timer on each update really buys anything and the code is now much cleaner.

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,697 @@
+		}
+	}
+	for index, skip := range skipDevice {
+		if skip {
```
Why not do this as `if dc.shouldSkipDevice(l)` in the loop above?

joamaki replied:
Because we need to first process all the link updates to figure out whether to skip or not, e.g. it might be created and then set as up. I'll add a comment!

joamaki replied:
This code is now gone. There's now only code to mark a device as "selected" (was viable, but selected seems like a better term).

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,697 @@
+// commits it to the state it returns true, if the batch is incomplete (e.g. route updates
+// for non-existing link), it returns false and the batch should be expanded with newer
+// updates and tried again.
+func (dc *devicesController) processBatch(init bool, batch map[int][]any) (ok bool) {
```
Nit: this could take `txn` instead of `init` and then init specific code could move into `initialize()`? I find bools in arguments hard to understand.

joamaki replied:
Good point, cleaned this up.

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,697 @@
+	l3DevSupported bool
+	ctx            context.Context
+	cancel         context.CancelFunc
+	links          map[int]netlink.Link
```
Seems like this is state which we need to keep in addition to what is in the statedb? This would benefit from documentation when that state needs to be invalidated, and why its necessary in the first place.

joamaki replied:
I'll clean this up by passing this state to `processBatch` to make it structurally clear that this state gets thrown away when netlink failure occurs and the whole thing restarts.

joamaki replied:
`links` is now gone and all state is in the devices and routes tables.

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,697 @@
+	name := link.Attrs().Name
+
+	// Do not consider devices that are down.
+	operState := link.Attrs().OperState
```
Doesn't this contradict the doc comment? I would've expected this in isViableDevice.

joamaki replied:
Ah crap. I initially had it in isViableDevice, but due to seeing the tests fail in different environments due to seeing "default" links like sit0 in the fresh namespace I moved it into shouldSkipDevice. But of course a link that's down can still get addresses added to it and if it later comes back up we would've completely missed those added addresses!

I'm going to move it back to isViableDevice and make sure that `shouldSkipDevice` doesn't change it's mind based on the state of the link. And will document this invariant.

joamaki replied:
I decided to significantly simplify this and merge `isViableDevice` and `shouldSkipDevice` into one. Now the `devices` table contains all system devices and those get marked as selected when they're in a ready state. This way it doesn't matter if address updates come before link updates for devices or if they go back and forth between accepted states.

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,697 @@
+		return true
+	}
+
+	if dc.filter.defined() {
```
The comment + not knowing what `defined()` does makes this a head scratcher for me. How does the filter know that this device has been specified by the user if we don't give it any parameters?

joamaki replied:
Doh, I chose a poor name for it. Point was to check whether a filter was defined (as in user had specified what devices should be used). Would `isEmpty()` or similar be better?
EDIT: Went with `nonEmpty() bool` and documented `deviceFilter`.

lmb commented on pkg/datapath/tables/route.go:
```diff
@@ -0,0 +1,120 @@
+	LinkIndex int
+
+	Scope uint8
+	Dst   *net.IPNet
```
Why doesn't `netip.Prefix` work?

joamaki replied:
It would. Just laziness in my part in going with the types coming from `netlink`. I'll switch this over to `netip`.

lmb reviewed (commented):
Phew! I can imagine that bashing this into shape has been quite a lot of work! I found this PR difficult to review. What would've helped me make quicker progress is a gradual refactor in multiple smaller commits, and as I mentioned maybe a code walk where you can explain the outline and the intricacies "in person". As it is I can't really figure out whether the new code has roughly the same semantics as the old one, so most of my comments are fairly generic, sorry. Is there someone who has worked on devices.go that we could pull in for additional review?

tommyp1ckles commented on pkg/datapath/tables/route.go:
```diff
@@ -0,0 +1,120 @@
+	"github.com/cilium/cilium/pkg/statedb"
+)
+
+type Route struct {
```
Q: Does defining a new route type with a subset of fields have something to do with how statedb works - couldn't we use the netlink.Route type (I see we use netlink.Link)

joamaki replied:
I'm trying to avoid exposing netlink here as it's becoming more and more likely we'll switch away from the unmaintained `netlink` library to something else.

tommyp1ckles commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,690 @@
+	}
+
+	if !dc.processBatch(true, batch) {
+		// This can only happen if the statedb rejects the commit.
```
seems like it can also happen if there is an updated related to a link that doesn't exist

joamaki replied:
This is no longer the case in the latest version of the code. Resolving.

tommyp1ckles commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,690 @@
+
+		name := link.Attrs().Name
+
+		d, _ := devicesWriter.First(tables.DeviceByIndex(index))
```
the interface comments say this can return an error if the query is malformed, the query here is pretty simple but this could perhaps lead to some confusing issues if its unclear whether it was a bad query or just an empty table.

joamaki replied:
Handling this explicitly now.

tommyp1ckles commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,690 @@
+}
+
+func hasGlobalRoute(devIndex int, routes statedb.TableReader[*tables.Route]) bool {
+	iter, _ := routes.Get(tables.RouteByLinkIndex(devIndex))
```
nit: should catch error and panic here ? otherwise this will crash on 616.

joamaki replied:
Done

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,653 @@
+		}
+
+		for _, u := range updates {
+			if d == nil {
```
Is this necessary even though there is an `if d == nil` at the end of the loop?

joamaki replied:
Point is to stop processing further updates for this link. I could instead add a break label and do `break updateLoop` on line 438, but this seems nicer.

EDIT: Ah, but now I'm likely making unfounded assumption here about ifindex not being reused too quickly. Will figure out what I can actually assume and fix the code accordingly. Wonder what fun races this thing has with regards to addresses and routes. Likely the correct way to do this (which is not easy with this netlink library) is to have a single netlink socket over which we get address, link and route updates.

EDIT2: At least locally I'm not seeing any reuse of ifindex. Not quite sure how paranoid this code should be. Thoughts?

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,653 @@
+
+// nonEmpty returns true if the filter has been defined
+// (i.e. user has specified --devices).
+func (lst deviceFilter) nonEmpty() bool {
```
Nice, thanks

lmb commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,653 @@
+
+// netlinkFuncs wraps the netlink subscribe functions into a simpler interface to facilitate
+// testing of the error handling paths.
+type netlinkFuncs struct {
```
Not saying you need to change this, but I'm curious: why a struct of func pointers instead of an interface?

joamaki replied:
Less boilerplate and makes it easy to override/wrap one of the methods to inject slightly different behavior in tests.

lmb reviewed (approved):
Thanks for your changes, it reads easier now.

tommyp1ckles commented on pkg/datapath/linux/devices_controller.go:
```diff
@@ -0,0 +1,653 @@
+		case <-ctx.Done():
+			stop()
+			return
+		case <-t.After(restartWaitDuration):
```
I'm guessing there's no need for a backoff here?

joamaki replied:
I removed the backoff in the end as it doesn't really bring much value, plus the logic for resetting the backoff would've been a bit involved (when would we even reset it?).

tommyp1ckles commented on pkg/datapath/tables/route.go:
```diff
@@ -0,0 +1,107 @@
+	Gw    netip.Addr
+}
+
+func (r *Route) DeepCopy() *Route {
```
Is this really a deepcopy? `r.Dst.{IP,Mask}`are still going to point to the underlying byte array of `r`.

joamaki replied:
They're now `netip.Addr` and `netip.Prefix` so this is fixed.

brb commented on daemon/cmd/daemon.go:
```diff
@@ -907,14 +902,16 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
+	var devices []string
+	if d.deviceManager != nil {
+		if _, err := d.deviceManager.Detect(params.Clientset.IsEnabled()); err != nil {
+			if option.Config.AreDevicesRequired() {
```
It'd be call if in the future each Cell could individually set a bit saying that the device detection is required.

joamaki replied:
I would get rid of this `AreDevicesRequired` stuff completely in the future and have each feature decide how to function or not without any devices.

tommyp1ckles reviewed (approved):
ok looks good on my end, thanks for bearing with me 😄
